src/generic/gridsel.cpp

```diff
@@ -975,222 +359,331 @@ void wxGridSelection::ClearSelection()
-    for ( n = 0; n < count; n++ )
-    {
-        wxGridCellCoords& coords1 = m_blockSelectionTopLeft[n];
-        wxGridCellCoords& coords2 = m_blockSelectionBottomRight[n];
-        wxCoord col1 = coords1.GetCol();
-        wxCoord col2 = coords2.GetCol();
+        wxGridBlockCoords& block = m_selection[n];
+        wxCoord col1 = block.GetLeftCol();
+        wxCoord col2 = block.GetRightCol();
 
         if ((size_t)col2 >= pos)
         {
             if (numCols > 0)
             {
                 // If rows inserted, increase row counter where necessary
-                coords2.SetCol(col2 + numCols);
+                block.SetRightCol(col2 + numCols);
                 if ((size_t)col1 >= pos)
-                    coords1.SetCol(col1 + numCols);
+                    block.SetLeftCol(col1 + numCols);
             }
             else if (numCols < 0)
             {
                 // If cols deleted ...
                 if ((size_t)col2 >= pos - numCols)
                 {
                     // ...either decrement col counter (if col still exists)...
-                    coords2.SetCol(col2 + numCols);
+                    block.SetRightCol(col2 + numCols);
                     if ( (size_t) col1 >= pos)
-                        coords1.SetCol( wxMax(col1 + numCols, (int)pos) );
+                        block.SetLeftCol( wxMax(col1 + numCols, (int)pos) );
 
                 }
                 else
                 {
                     if ((size_t)col1 >= pos)
                     {
                         // ...or remove the attribute
-                        m_blockSelectionTopLeft.RemoveAt(n);
-                        m_blockSelectionBottomRight.RemoveAt(n);
+                        m_selection.erase(m_selection.begin() + n);
                         n--;
                         count--;
                     }
                     else
-                        coords2.SetCol(pos);
+                        block.SetRightCol(pos);
                 }
             }
         }
     }
+}
 
-    count = m_colSelection.GetCount();
-    for ( n = 0; n < count; n++ )
+bool wxGridSelection::ExtendOrCreateCurrentBlock(const wxGridCellCoords& blockStart,
+                                                 const wxGridCellCoords& blockEnd,
+                                                 const wxKeyboardState& kbd)
+{
+    if ( m_selection.empty() )
+    {
+        SelectBlock(blockStart, blockEnd);
+        return true;
+    }
+
+    wxGridBlockCoords& block = *m_selection.rbegin();
+    wxGridBlockCoords newBlock = block;
+
+    bool editBlock = false;
+
+    if ( blockEnd.GetRow() != -1 )
+    {
+        if ( newBlock.GetTopRow() == blockStart.GetRow() )
+        {
+            newBlock.SetBottomRow(blockEnd.GetRow());
+            editBlock = true;
+        }
+        else if ( newBlock.GetBottomRow() == blockStart.GetRow() )
+        {
+            newBlock.SetTopRow(blockEnd.GetRow());
+            editBlock = true;
+        }
```

Comment:
Just FYI, all this doesn't work at all for column selection mode (and, of course, the reversed version below doesn't work for the much more often used row selection mode), it results in having non-column selection blocks as it adjusts rows to be different from the 0/last.

Reply:
Moreover, the selection misbehaves, even in cell selection mode, when neither of the 2 conditions are satisfied. There really must be something in the (currently missing) `else` branch here.

Reply:
For the reference, this "something" is the code to expand the current block to the newly specified block in this direction. This is required to allow Shift-clicking on the header to extend the existing selection correctly and only expanding, but not shrinking, the current block is required for Shift-arrows to preserve the full line selection afterwards.